src/lib/core/style/focus-classes.ts

```diff
@@ -39,21 +56,37 @@ export class FocusOriginMonitor {
   }
 
   /** Handles focus events on a registered element. */
-  private _onFocus(element: Element, renderer: Renderer) {
+  private _onFocus(element: Element, renderer: Renderer, subject: Subject<FocusOrigin>) {
+    // If we couldn't detect a cause for the focus event, it's due to one of two reasons:
+    // 1) The window has just regained focus, in which case we want to restore the focused state of
+    //    the element from before the window blurred.
+    // 2) The element was programmatically focused, in which case we should mark the origin as
+    //    'program'.
+    if (!this._origin) {
+      if (this._windowFocused && this._lastFocused.has(element)) {
+        this._origin = this._lastFocused.get(element);
+      } else {
+        this._origin = 'program';
+      }
+    }
+
     renderer.setElementClass(element, 'cdk-focused', true);
     renderer.setElementClass(element, 'cdk-keyboard-focused', this._origin == 'keyboard');
     renderer.setElementClass(element, 'cdk-mouse-focused', this._origin == 'mouse');
-    renderer.setElementClass(element, 'cdk-program-focused',
-        !this._origin || this._origin == 'program');
+    renderer.setElementClass(element, 'cdk-program-focused', this._origin == 'program');
+
+    subject.next(this._origin);
+    this._lastFocused = new WeakMap().set(element, this._origin);
```

Comment:
I'm trying to understand why we need a `WeakMap` when the the WeakMap always contains one element at most.

I think we can assume that the browser always returns focus to the last focused element and multiple elements can't be focused.